File: src/diff.cpp
#include "diff.hpp"

#include "keyword.hpp"
#include "tokenize.hpp"
#include "vector_math.hpp"

#include <algorithm>
#include <cctype>
#include <string>
#include <unordered_map>

namespace mnemon::search_engine {

// Phrases that flip “similar text” into a semantic conflict (English + Chinese), aligned with Go diff heuristics.
static const char* kNeg[] = {"not",           "no longer",     "don't",     "doesn't",    "never",
                             "switched from", "instead of",    "rather than", "replaced", "deprecated",
                             "不",            "没有",          "不再",      "放弃",       "替换",
                             "取消"};

// Thresholds: low sim → ADD; negation → CONFLICT; very high → DUPLICATE; else UPDATE.
static DiffSuggestion classify_suggestion(double similarity, std::string_view new_text, std::string_view old_text) {
  if (similarity < 0.5) {
    return DiffSuggestion::Add;
  }
  std::string nl(new_text);
  std::string ol(old_text);
  for (auto& c : nl) {
    c = static_cast<char>(std::tolower(static_cast<unsigned char>(c)));
  }
  for (auto& c : ol) {
    c = static_cast<char>(std::tolower(static_cast<unsigned char>(c)));
  }
  for (const char* neg : kNeg) {
    if (nl.find(neg) != std::string::npos || ol.find(neg) != std::string::npos) {
      return DiffSuggestion::Conflict;
    }
  }
  if (similarity > 0.9) {
    return DiffSuggestion::Duplicate;
  }
  return DiffSuggestion::Update;
}

std::string diff_suggestion_str(DiffSuggestion s) {
  switch (s) {
  case DiffSuggestion::Add:
    return "ADD";
  case DiffSuggestion::Duplicate:
    return "DUPLICATE";
  case DiffSuggestion::Conflict:
    return "CONFLICT";
  case DiffSuggestion::Update:
    return "UPDATE";
  }
  return "ADD";
}

// 1) Keyword candidates, blend token vs cosine sim when embeddings exist.
// 2) Second pass: high-cosine pairs missed by keyword search (embedding-only near-dupes).
DiffResult diff_insights(const std::vector<Insight>& insights, std::string_view new_content, const DiffOptions& opts) {
  int lim = opts.limit > 0 ? opts.limit : 5;
  auto candidates = keyword_search(insights, new_content, lim);
  std::unordered_map<std::string, std::vector<double>> embed_map;
  for (const auto& e : opts.existing_embed) {
    embed_map[e.id] = e.embedding;
  }

  std::vector<DiffMatch> matches;
  for (const auto& c : candidates) {
    double token_sim = jaccard_similarity(new_content, c.insight.content);
    double cosine_sim = 0;
    if (!opts.new_embedding.empty()) {
      auto it = embed_map.find(c.insight.id);
      if (it != embed_map.end() && !it->second.empty()) {
        cosine_sim = mnemon::cosine_similarity(opts.new_embedding, it->second);
      }
    }
    double similarity = token_sim;
    // Combined similarity: cosine only contributes when above 0.85.
    // Below that, same-domain content clusters around 0.70–0.84 and produces false UPDATE.
    if (cosine_sim >= 0.85 && cosine_sim > similarity) {
      similarity = cosine_sim;
    }
    auto sug = classify_suggestion(similarity, new_content, c.insight.content);
    matches.push_back(DiffMatch{c.insight.id, c.insight.content, token_sim, cosine_sim, similarity, sug});
  }

  if (!opts.new_embedding.empty() && !opts.existing_embed.empty()) {
    std::unordered_map<std::string, bool> seen;
    for (const auto& m : matches) {
      seen[m.id] = true;
    }
    struct P {
      std::string id;
      double sim;
    };
    std::vector<P> top_cos;
    for (const auto& ei : opts.existing_embed) {
      if (seen[ei.id]) {
        continue;
      }
      double cs = mnemon::cosine_similarity(opts.new_embedding, ei.embedding);
      if (cs >= 0.85) {
        top_cos.push_back({ei.id, cs});
      }
    }
    std::sort(top_cos.begin(), top_cos.end(), [](const P& a, const P& b) { return a.sim > b.sim; });
    if (static_cast<int>(top_cos.size()) > lim) {
      top_cos.resize(static_cast<size_t>(lim));
    }
    std::unordered_map<std::string, Insight> id_to_ins;
    for (const auto& ins : insights) {
      id_to_ins[ins.id] = ins;
    }
    for (const auto& cp : top_cos) {
      auto it = id_to_ins.find(cp.id);
      if (it == id_to_ins.end()) {
        continue;
      }
      double token_sim = jaccard_similarity(new_content, it->second.content);
      double similarity = token_sim;
      if (cp.sim >= 0.85 && cp.sim > similarity) {
        similarity = cp.sim;
      }
      auto sug = classify_suggestion(similarity, new_content, it->second.content);
      if (sug != DiffSuggestion::Add) {
        matches.push_back(
            DiffMatch{it->second.id, it->second.content, token_sim, cp.sim, similarity, sug});
      }
    }
  }

  // Overall suggestion: any DUPLICATE among candidates wins (stop remember-as-new).
  DiffSuggestion overall = DiffSuggestion::Add;
  if (!matches.empty()) {
    overall = matches[0].suggestion;
    for (const auto& m : matches) {
      if (m.suggestion == DiffSuggestion::Duplicate) {
        overall = DiffSuggestion::Duplicate;
        break;
      }
    }
  }
  return DiffResult{overall, std::move(matches)};
}

} // namespace mnemon::search_engine

File: mnemon/src/engine/diff.ts
import { DUPLICATE_MAX_LENGTH_RATIO, DUPLICATE_TOKEN_SIMILARITY } from "./constants.js";
import { codePointLength } from "./normalize.js";
import { cosineSimilarity, jaccardTokenSimilarity, setsEqual, symmetricTokenSimilarity } from "./similarity.js";
import { tokenize } from "./tokenize.js";

export interface DuplicateCandidate {
  id: string;
  content: string;
  tokenSimilarity: number;
  cosineSimilarity: number;
}

const NEGATION = new Set(["not", "no", "never"]);

function negationMarks(text: string): string {
  const marks: string[] = [];
  for (const word of text.toLowerCase().match(/\p{L}+/gu) ?? []) {
    if (NEGATION.has(word)) {
      marks.push(word);
    }
  }
  return marks.sort().join(",");
}

export function isSafeDuplicate(newContent: string, existingContent: string, cosine = 0): boolean {
  const newTokens = tokenize(newContent);
  const existingTokens = tokenize(existingContent);
  const tokenSim = symmetricTokenSimilarity(newTokens, existingTokens);
  if (tokenSim <= DUPLICATE_TOKEN_SIMILARITY) {
    return false;
  }
  if (codePointLength(newContent) > codePointLength(existingContent) * DUPLICATE_MAX_LENGTH_RATIO) {
    return false;
  }
  if (negationMarks(newContent) !== negationMarks(existingContent)) {
    return false;
  }
  void cosine;
  return setsEqual(newTokens, existingTokens);
}

export function scoreDuplicateCandidate(
  newContent: string,
  existingContent: string,
  newEmbedding?: readonly number[],
  existingEmbedding?: readonly number[],
): { tokenSimilarity: number; cosineSimilarity: number } {
  const tokenSimilarity = symmetricTokenSimilarity(tokenize(newContent), tokenize(existingContent));
  const cosine =
    newEmbedding && existingEmbedding ? cosineSimilarity(newEmbedding, existingEmbedding) : 0;
  return { tokenSimilarity, cosineSimilarity: cosine };
}

export function classifySafeDuplicate(
  newContent: string,
  candidates: readonly DuplicateCandidate[],
): DuplicateCandidate | undefined {
  for (const candidate of candidates) {
    if (isSafeDuplicate(newContent, candidate.content, candidate.cosineSimilarity)) {
      return candidate;
    }
  }
  return undefined;
}

export type DiffSuggestion = "ADD" | "DUPLICATE" | "CONFLICT" | "UPDATE";

export interface DiffMatch {
  id: string;
  content: string;
  tokenSimilarity: number;
  cosineSimilarity: number;
  similarity: number;
  suggestion: DiffSuggestion;
}

export interface DiffResult {
  suggestion: DiffSuggestion;
  matches: DiffMatch[];
}

/** Same phrase list as Go. Bare "not" is excluded so scientific text does not look like CONFLICT. */
export const DIFF_NEGATION_PHRASES = [
  "no longer",
  "don't",
  "doesn't",
  "never",
  "switched from",
  "instead of",
  "rather than",
  "replaced",
  "deprecated",
  "不再",
  "放弃",
  "替换",
  "取消",
] as const;

function isExtension(newText: string, existingText: string): boolean {
  const n = codePointLength(newText);
  const e = codePointLength(existingText);
  return n > e + Math.floor(e / 4);
}

function hasConflictPhrase(newText: string, existingText: string): boolean {
  const newer = newText.toLowerCase();
  const older = existingText.toLowerCase();
  return DIFF_NEGATION_PHRASES.some((phrase) => newer.includes(phrase) || older.includes(phrase));
}

export function classifySuggestion(
  tokenSim: number,
  similarity: number,
  newText: string,
  existingText: string,
): DiffSuggestion {
  if (similarity < 0.5) {
    return "ADD";
  }
  const extension = isExtension(newText, existingText);
  if (tokenSim > 0.9 && !extension) {
    return "DUPLICATE";
  }
  if (similarity >= 0.7 && hasConflictPhrase(newText, existingText)) {
    return "CONFLICT";
  }
  if (similarity > 0.9 && !extension) {
    return "DUPLICATE";
  }
  return "UPDATE";
}

function combinedSimilarity(tokenSim: number, cosineSim: number): number {
  if (cosineSim >= 0.85 && cosineSim > tokenSim) {
    return cosineSim;
  }
  return tokenSim;
}

export function classifyDiff(
  newContent: string,
  candidates: readonly { id: string; content: string; cosineSimilarity?: number }[],
): DiffResult {
  const newTokens = tokenize(newContent);
  const matches: DiffMatch[] = candidates.map((candidate) => {
    const tokenSimilarity = jaccardTokenSimilarity(newTokens, tokenize(candidate.content));
    const cosineSimilarity = candidate.cosineSimilarity ?? 0;
    const similarity = combinedSimilarity(tokenSimilarity, cosineSimilarity);
    return {
      id: candidate.id,
      content: candidate.content,
      tokenSimilarity,
      cosineSimilarity,
      similarity,
      suggestion: classifySuggestion(tokenSimilarity, similarity, newContent, candidate.content),
    };
  });
  matches.sort((a, b) => b.similarity - a.similarity || a.id.localeCompare(b.id));
  let suggestion: DiffSuggestion = matches[0]?.suggestion ?? "ADD";
  if (matches.some((m) => m.suggestion === "DUPLICATE")) {
    suggestion = "DUPLICATE";
  }
  return { suggestion, matches };
}
